File: projects/7/VMTranslator/src/main/java/com/starki/VMTranslator.java
package com.starki;

public class VMTranslator {
    public static void main(String[] args) {
        // 确保传入的参数是有效的文件路径
        if (args.length != 1) {
            System.err.println("Usage: java com.starki.VMTranslator <VM file path>");
            return;
        }

        String filePath = args[0]; // 获取命令行参数中的文件路径

        if (!filePath.endsWith(".vm"))// 检查文件路径是否以.vm结尾
        {
            System.err.println("Error: Invalid VM file - " + filePath);
            return;
        }

        getCodeWrite(filePath);
    }

    private static void getCodeWrite(String filePath) {
        Parser parser = new Parser(filePath);// 打开并解析输入文件
        CodeWrite codeWriter = new CodeWrite(filePath);

        // 逐条命令进行翻译
        while (parser.hasMoreCommands()) {
            parser.advance(); // 获取下一条命令

            String commandType = parser.commandType(); // 获取当前命令的类型

            if (commandType.equals("C_ARITHMETIC")) {
                // 如果是算术命令，写入相应的汇编代码
                codeWriter.writeArithmetic(parser.arg1());
            } else if (commandType.equals("C_PUSH") || commandType.equals("C_POP")) {
                // 如果是推送或弹出命令，写入相应的汇编代码
                codeWriter.writePushPop(commandType, parser.arg1(), parser.arg2());
            }
        }

        // 完成所有命令翻译后，关闭文件流
        codeWriter.close();
        System.out.println("Translation completed successfully.");
    }
}

File: nand2TetrisCpp/VMTranslatorCpp/src/VMTranslator.cpp
#include "CodeWriter.h"
#include "Parser.h"
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <vector>

using namespace std;

static bool isVM(const std::filesystem::path &p) {
  return p.has_extension() &&
         (p.extension() == ".vm" || p.extension() == ".VM");
}

int main(int argc, char **argv) {
  if (argc != 2) {
    cerr << "Usage: vmtranslator <input .vm file or directory>\n";
    return 1;
  }
  std::filesystem::path in(argv[1]);

  if (!std::filesystem::exists(in)) {
    cerr << "Input path does not exist: " << in << "\n";
    return 1;
  }

  vector<std::filesystem::path> files;
  std::filesystem::path out_path;

  if (std::filesystem::is_directory(in)) {
    for (auto &entry : std::filesystem::directory_iterator(in)) {
      if (entry.is_regular_file() && isVM(entry.path())) {
        files.push_back(entry.path());
      }
    }
    sort(files.begin(), files.end());
    if (files.empty()) {
      cerr << "No .vm files found in directory: " << in << "\n";
      return 1;
    }
    out_path = in / (in.filename().string() + ".asm");
  } else {
    if (!isVM(in)) {
      cerr << "Input file must have .vm extension\n";
      return 1;
    }
    files.push_back(in);
    out_path = in.parent_path() / (in.stem().string() + ".asm");
  }

  try {
    CodeWriter writer(out_path.string());
    /*todo*/ // 目录模式是否需要 bootstrap?

    for (const auto &f : files) {
      Parser parser(f.string());
      writer.setFileName(parser.moduleName());

      while (parser.hasMoreCommands()) {
        auto type = parser.commandType();
        if (type == "C_ARITHMETIC") {
          writer.writeArithmetic(parser.arg1());
        } else {
          /*todo*/ // 其余分派：push/pop/label/goto/if/function/call/return
        }
        parser.advance();
      }
    }
    writer.close();
    cout << "Wrote: " << out_path << "\n";
  } catch (const std::exception &e) {
    cerr << "Error: " << e.what() << "\n";
    return 2;
  }
  return 0;
}
